atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -1808,18 +1808,26 @@ public void commit(TransactionService transactionService) {
     }
 
     private void checkConstraints() {
-        List<String> violations = new ArrayList<>();
-        for (Map.Entry<TableReference, ConstraintCheckable> entry : constraintsByTableName.entrySet()) {
-            SortedMap<Cell, byte[]> sortedMap = writesByTable.get(entry.getKey());
-            if (sortedMap != null) {
-                violations.addAll(entry.getValue().findConstraintFailures(sortedMap, this, constraintCheckingMode));
-            }
+        if (constraintsByTableName.isEmpty() || writesByTable.isEmpty()) {
```

Comment:
this will traverse both of these `ConcurrentHashMap`'s buckets to get a size() count, but will not allocate. In many cases `constraintsByTableName` will be empty when there are no constraints and short circuit, and in the case of read only transactions, `writesByTable` will be empty, so there is no work to check constraints.

Reply:
Sadly, `constraintsByTableName will be empty when there are no constraints` is not strictly true - empty `constraintsByTable` implies no constraints, but no constraints does not imply an empty `constraintsByTable`. In fact, given the current implementation of SnapshotTransaction, there should be an entry for every unique table written to in the transaction, even if the constraint-providing method just returns the empty list. (This is because `SnapshotTransaction.useTable`, which is called when a table has been written to, creates a new entry in the `constraintsByTableName` map).

In summary, it would be sufficient to check just when `writesByTable` is empty, since entries in `writesByTable` implies at least one entry in `constraintsByTable`. This would only short-circuit for read-only transactions, and not when a table has no constraints.

Let me know if you'd like to take a shot at implementing no constraints => empty `constraintsByTable`, otherwise I can approve with the main read-only transaction improvement :)

Reply:
good call, when there are no constraints for a table, `findConstraintFailures` will return an empty list, so I've reworked the stream below to filter out when there are no constraints and avoid the flatMap stream allocation for those cases.

Reply:
actually going to just revert back to the for loop for simplicity and minimizing churn. The `List<String> violations = new ArrayList<>();` will be a cheap allocation and almost always empty assuming there are no constraint violations.

Reply:
Thanks! One other thing I'm going to note for posterity is that if any ConstraintCheckable had side effects _and_ relied  on them happening on read transactions (meaning that incorrectly marked table as used), then this does break that. However, 1) I couldn't find any internal uses of that and 2) should not be marking a table as used when there are only reads, therefore I'm okay with this.